device/src/storage.rs

```diff
@@ -79,55 +81,88 @@ impl DeviceStorage {
     }
 
+    /// The layout of the entries are word aligned and length prefixed. So the first entry has a
+    /// four byte little endian entry length (in words), and then the main body of the item bincode
+    /// encoded. Bincode doesn't care if it doesn't use all the bytes so there's no need to know
+    /// exactly where it ends.
     pub fn append(
         &mut self,
         changes: impl IntoIterator<Item = Change>,
-    ) -> Result<(), bincode::error::EncodeError> {
+    ) -> Result<(), FlashStorageError> {
         for change in changes.into_iter() {
-            bincode::encode_into_writer(
-                change,
-                BufWriter(&mut self.write_buffer),
-                bincode::config::standard(),
-            )?;
-        }
-        if self.write_buffer.is_empty() {
-            return Ok(());
+            self.write_buffer.clear();
+            // placeholder for the length
+            self.write_buffer
+                .extend(core::iter::repeat_n(0x00, WORD_SIZE));
+            bincode::encode_into_writer(change, BufWriter(&mut self.write_buffer), BINCODE_CONFIG)
+                .expect("can bincode encode type to buffer");
+
+            let padding_needed = (WORD_SIZE - (self.write_buffer.len() % WORD_SIZE)) % WORD_SIZE;
+
+            self.write_buffer
+                .extend(core::iter::repeat_n(0xff, padding_needed));
+            assert_eq!(self.write_buffer.len() % WORD_SIZE, 0);
+
+            let word_length: u32 = (self.write_buffer.len() / WORD_SIZE)
+                .try_into()
+                .expect("flash item too large");
+
+            self.write_buffer[0..WORD_SIZE].copy_from_slice(
+                (word_length - 1/* don't include the length word */)
+                    .to_le_bytes()
+                    .as_slice(),
+            );
+            let byte_pos = (self.word_pos * WORD_SIZE) as u32;
+            nor_flash::NorFlash::write(&mut self.flash, byte_pos, &self.write_buffer[..])?;
+            self.word_pos += word_length as usize;
+            self.write_buffer.clear();
         }
-        self.flash
-            .write(self.pos, &self.write_buffer[..])
-            .map_err(|e| {
-                bincode::error::EncodeError::OtherString(format!("flash write error: {:?}", e))
-            })?;
-        self.pos += self.write_buffer.len() as u32;
-        self.write_buffer.clear();
 
         Ok(())
     }
 
     pub fn iter(&mut self) -> impl Iterator<Item = Change> + '_ {
-        self.pos = DATA_START;
         core::iter::from_fn(move || {
-            let pos_before_read = self.pos;
-            match bincode::decode_from_reader(&mut *self, bincode::config::standard()) {
-                Ok(change) => Some(change),
-                Err(_) => {
-                    self.pos = pos_before_read;
-                    None
-                }
+            let mut length_buf = [0u8; WORD_SIZE];
+            let byte_pos = self.word_pos * WORD_SIZE;
```

Comment:
Here we're only reading forward from `self.word_pos`, instead of from `DATA_START`.
Maybe intended since we only iter upon initial boot, making a note because it might be unexpected behaviour if we were to ever iter through the flash elsewhere in the code.